src/kernel.cr

```diff
@@ -160,12 +165,24 @@ def at_exit(&handler : Int32 ->) : Nil
   AtExitHandlers.add(handler)
 end
 
+class NestedExitError < Exception
+  getter status_code : Int32
+
+  def initialize(@status_code)
+  end
+end
+
 # Terminates execution immediately, returning the given status code
 # to the invoking environment.
 #
 # Registered `at_exit` procs are executed.
 def exit(status = 0) : NoReturn
-  AtExitHandlers.run status
+  if AtExitHandlers.running?
+    raise NestedExitError.new status
```

Comment:
it shouldn't be possible to rescue a nested exit, and it shouldn't use stack unwinding. Please create a state machine where calling `exit` simply resumes where the previous exit call left off.